### ISSUE TITLE
ASCollectionNode Bugs on Master Branch

### ISSUE THREAD

**Issue by QuynhNguyen** [facebook/AsyncDisplayKit#3074]
<img width="535" alt="screen shot 2017-02-24 at 1 01 25 am" src="https://cloud.githubusercontent.com/assets/918890/23297106/020b8b98-fa2d-11e6-984f-cef8565b3b2e.png">

On version 2.2 and under, the search bar would display properly below the status bar but on `master` the status bar would overlap it a bit. The following is my collection node hierarchy

- Section 1
    - Supplementary view header which contains the `UISearchBar`
- Section 2
    - cell1
    - cell2

**Comment by nguyenhuy** [comment]
Hi, thanks for reporting. Could you please help us to bisect recent commits?

**Comment by appleguy** [comment]
@QuynhNguyen Yes, thanks for filing this.  I was about to file an issue myself too, as I'm encountering something very similar with supplementary nodes (in my case, crashing).

@nguyenhuy Fortunately this can be reproduced right away with examples/CustomCollectionView.  I suggest you have my PR applied just to make sure you don't hit the unnecessary assertions at a stage earlier than the actual crash: (this is now in master, so just pull to latest) https://github.com/facebook/AsyncDisplayKit/pull/3079

The attached .txt has some debug logs that I found helpful.  Here's the issue:

- The data source is queried correctly, and does return supplementary nodes for everything that it should.
- These are put into an ASMutableElementMap and I believe this does make it to _pendingMap (but it's a bit hard to tell since it is copied)
- However, at the time of the crash, the _visibleMap object that is being queried by UIKit does not have _any_ supplementary elements in it...even though we've reported that some exist to UIKit.

I think it is possible that the async sizing pass for the cells somehow doesn't gate the collection becoming aware of the supplementary nodes earlier, before the sizing pass is done.

@nguyenhuy @Adlai-Holler thank you so much for all the work on ASDataController!  It is very architecturally impressive, and truly represents a successful new era of one of the most crucial components in the framework.  These bugs are likely to be small fixes.

[SupplementaryMapLogging.txt](https://github.com/facebook/AsyncDisplayKit/files/801686/SupplementaryMapLogging.txt)
